Add 304 tests for CORS

mnot opened the pull request:
... replacement for #1400, more changes to come.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/1740)

<!-- Reviewable:end -->

mnot committed:
Add 304 tests for CORS.

CORS doesn't specify what to do with a 304 response, but
HTTP specifies that its headers should be combined with
the stored response -- i.e., a 304 should be under the
same CORS policy as the original 200 (unless it changes it).

This is testing to make sure that this is the case, since
many servers (including intermediaries) do NOT add CORS
headers (and some can't -- e.g., intermediary caches).

Future tests might include things like updates to CORS in
those 304s (fancy!).

mnot committed:
Update to check Access-Control-Expose-Headers on 304

mnot committed:
Remove trailing whitespace b/c lint is SO GRUMPY

mnot committed:
check to see if a 304 can change Access-Control-Allow-Origin

mnot committed:
oops, req_num

mnot committed:
client4, not client3

mnot committed:
go async

mnot committed:
bow before the infinite wisdom of ./lint

mnot committed:
Remove superfluous 304Response:

as per Odin

mnot committed:
Factor out state header, disable checks

mnot committed:
Merge branch 'master' of https://github.com/w3c/web-platform-tests into 304-1

annevk commented:
@mnot by the way, I'm happy to make the changes to the specification if you could give an outline of what you think needs to be done.

mnot commented:
@annevk I think the changes would be in 4.3 HTTP network or cache fetch, around step 7. 

That assumes that a cache entry, when present, can only be "complete" or "incomplete", but there's another axis -- "fresh" or "stale".

What is specified handles "fresh", but if it's stale, the request still needs to be made, but as a conditional request, and the response if a 304 needs to be combined with the cached response, as per <http://httpwg.github.io/specs/rfc7234.html#freshening.responses>.

That would handle *this* issue. There's a lot more that isn't covered, such as:

* Serving stale responses when the network is down - <http://httpwg.github.io/specs/rfc7234.html#serving.stale.responses>
* Handling Vary - <http://httpwg.github.io/specs/rfc7234.html#caching.negotiated.responses>
* Cache invalidation - <http://httpwg.github.io/specs/rfc7234.html#invalidation>
* Request cache-control - <http://httpwg.github.io/specs/rfc7234.html#cache-request-directive>

Etc. Ideally you wouldn't need to re-specify all of caching, but you do need to accommodate the major features if you're going to go to the level of detail that you have.

Does that make sense generally? If so, I'll start a pull.

Also, in step 8, what are "resume headers"?

annevk commented:
That sounds reasonable. I too would like to avoid respecifying all that, but I'm not sure that is possible and still provide detailed control over it. "resume headers" I guess is a placeholder name for `Range` and friends.

mnot commented:
@annevk does it make sense to take another look at this now?

annevk commented on cors/304.htm:
```diff
@@ -0,0 +1,105 @@
+  console.log('starting request ' + id)
+  var client = new XMLHttpRequest()
+  client.error_caught = false
+  try {
```
Asynchronous XMLHttpRequest does not throw (unless you do method invocation incorrectly) so `try` is not needed.

mnot replied:
Fixed, thanks.

annevk commented on cors/304.htm:
```diff
@@ -0,0 +1,105 @@
+    if (client.readyState == client.DONE) {
+      console.log('finished request ' + id)
+      t.step(function() {
+        assert_true(client.status != 299, "req " + id + " server says: " + client.responseText)
```
What should status be? Wouldn't it be better to test for that?

mnot replied:
This is just testing for a server-specific error, denoted by 299 (yes, I realise how hypocritical that makes me...). The actual status code expected depends upon the specific test being run.

annevk commented on cors/304.htm:
```diff
@@ -0,0 +1,105 @@
+      assert_equals(client1.getResponseHeader(state_header), id1, "1st response didn't come from server;")
+    })
+
+    setTimeout(function() {
```
Why do the second request from a timeout rather than the ready of the first request?

mnot replied:
We're testing cache behaviour, so we need to have a clear separation (in time) between the requests, so that if something is happening async (e.g., storing the response and making it available), we don't get and conflicts. Yes, there's potentially a race here, but I don't think that's avoidable.

annevk replied:
Maybe add a comment to that effect? In particular it seems you want to avoid doing the second request while the first request's response might still be written to the cache. In that case it still seems better to start the timer after you received the response from the first request, no?

mnot replied:
Comment added. It's already starting the timer after the first response is DONE.

annevk replied:
Instead of setTimeout, use `t.step_timeout`. That makes the test more reliable.

annevk commented on cors/304.htm:
```diff
@@ -0,0 +1,105 @@
+
+async_test(function(t) {
+  two_reqs('1', '2', true, t, function(client1, client2) {
+    assert_equals(client1.response, client2.response, "responses were different;")
```
This is already checked above, no?

mnot replied:
Der, yes. Fixed.

annevk commented on cors/304.htm:
```diff
@@ -0,0 +1,105 @@
+
+async_test(function(t) {
+  two_reqs('7', '8', false, t, function(client1, client2) {
+    if (! client2.error_caught) {
```
I don't think this works. Why would it have thrown?

mnot replied:
Not sure what I was thinking here (likely a leftover from old code), but fixed now.

annevk commented:
@mnot do you still plan on reviving this?

mnot commented:
Yes, will page it back in tomorrow.

mnot committed:
Merge branch 'master' of https://github.com/w3c/web-platform-tests into 304-1

mnot committed:
don't try to catch on an async xhr

mnot committed:
add some comments

mnot committed:
remove spurious assert

mnot committed:
remove error_caught remnants

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/204088740)

# Firefox (nightly channel) #
Testing web-platform-tests at revision db2a8344561a28a6bada89ba80edb8e88a53ab8b
Using browser at version BuildID 20170221110300; SourceStamp d84beb192e57e26846c82d3df3599381f4663792
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/1740/cors/304.htm">/cors/304.htm</a></summary>


|                                 Subtest                                 | Results | Messages |
|-------------------------------------------------------------------------|---------|----------|
|                                                                         | OK      |          |
| `A 304 response with no CORS headers inherits from the stored response` | PASS    |          |
| `A 304 can expand Access-Control-Expose-Headers`                        | PASS    |          |
| `A 304 can contract Access-Control-Expose-Headers`                      | PASS    |          |
| `A 304 can change Access-Control-Allow-Origin`                          | PASS    |          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/204088741)

# Chrome (unstable channel) #
Testing web-platform-tests at revision db2a8344561a28a6bada89ba80edb8e88a53ab8b
Using browser at version 58.0.3013.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/1740/cors/304.htm">/cors/304.htm</a></summary>


|                                 Subtest                                 | Results | Messages |
|-------------------------------------------------------------------------|---------|----------|
|                                                                         | OK      |          |
| `A 304 response with no CORS headers inherits from the stored response` | PASS    |          |
| `A 304 can expand Access-Control-Expose-Headers`                        | PASS    |          |
| `A 304 can contract Access-Control-Expose-Headers`                      | PASS    |          |
| `A 304 can change Access-Control-Allow-Origin`                          | PASS    |          |

</details>
</details>

mnot committed:
let the caller decide how to test this

mnot commented:
I addressed the comments above. Chrome and Firefox stable both pass all tests; Safari seems to have had two regressions since I wrote this way back when.

annevk commented:
You need to remove console logging from the tests. That's why they're failing.

annevk commented:
Also asking @youennf to take a look since Safari is failing.

mnot committed:
remove console logging

mnot committed:
use t.step_timeout

as per anne

mnot committed:
the linter doesn't even liked commented out logging.

annevk commented:
w3c-test:mirror

annevk commented:
w3c-test:mirror

annevk commented:
@mnot if Fx is passing all tests we're not testing https://bugzilla.mozilla.org/show_bug.cgi?id=1311566. Would be good to test that too, no? Can you also file a bug against Safari at https://bugs.webkit.org/enter_bug.cgi?product=WebKit (component HTML DOM I suppose)?

youennf commented:
Nice tests!
Not sure why WebKit is failing.
If you have time to file a bug in WebKit bugzilla...

sideshowbarker commented:
w3c-test:mirror

sideshowbarker commented:
https://w3c-test.org/submissions/1740/cors/

mnot commented:
@annevk yeah I realised that in the shower this morning :) Will get something in there.

mnot commented:
Actually, the first test should catch that, because the 304 response doesn't include `ACAO`.

annevk commented:
@mnot I guess the scenario you're not testing at the moment is making a conditional request with your own state. That is, you setting your own If-None-Match headers, but I think in that case we still want the 304 to have CORS headers.

Anyway, that can be follow up I think, I'll land this later today.

annevk commented:
It seems the Safari issues are fixed in Safari TP.

annevk commented:
(I wrote tests for the scenario I mentioned above, but all browsers still pass.)
